In [23]:
%%capture
%load_ext autoreload
%autoreload 2

%pip install -r requirements.txt

In [1]:
import pandas as pd

d:\anaconda3\envs\ame{{environment}}\Lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


## Vuelos

In [2]:
flights = pd.read_parquet("../resources/flights.parquet")
flights.head()

Flight_ID Aeronave DepartureStation ArrivalStation  \
0  ab954014077430bd842cfa305a55c0f8   XA-VBY               AT             AZ   
1  efd86c996035dacdca7a0ccb2560dda1   XA-VIX               BM             AV   
2  6cfa1bbaa44f08fc7d3061f034a6a5ce   XA-VBV               AW             AS   
3  dd0fad3248951d2f71d63e6279aeaa4b   XA-VBW               AW             AS   
4  d0987ee648eea254063bfe2b39571b67   XA-VAP               BA             AB   

      Destination_Type        Origin_Type                  STD  \
0    Ciudad Fronteriza   Ciudad Principal  2023-10-19 11:40:00   
1  MX Amigos y Familia  Ciudad Fronteriza  2023-07-03 00:55:00   
2  MX Amigos y Familia   Ciudad Principal  2024-02-16 17:10:00   
3  MX Amigos y Familia   Ciudad Principal  2023-06-26 15:15:00   
4                Playa   Ciudad Principal  2023-02-10 08:40:00   

                   STA  Capacity  Passengers  Bookings  
0  2023-10-19 14:25:00       240       229.0     157.0  
1  2023-07-03 04:55:00       186       197.0     109.0  
2  2024-02-16 17:55:00       220         NaN       NaN  
3  2023-06-26 15:55:00       220       200.0     142.0  
4  2023-02-10 09:50:00       186       162.0      90.0

Escoger los datos de enero 2024

In [3]:
flights["STD"] = pd.to_datetime(flights["STD"])
flights["STA"] = pd.to_datetime(flights["STA"])

In [4]:
#Show rows from before january 2024
flights = flights[flights["STD"] < "2024-01-01"]

In [5]:
# Registros nulos para cada campo
flights.isna().sum()

Flight_ID             0
Aeronave            502
DepartureStation      1
ArrivalStation        1
Destination_Type      1
Origin_Type           1
STD                   0
STA                   0
Capacity              0
Passengers            0
Bookings              0
dtype: int64

In [9]:
# Borrar los datos que no tienen Aeronave
flights = flights.dropna(subset=["Aeronave","DepartureStation","ArrivalStation","Destination_Type","Origin_Type"])

In [10]:

flights.isna().sum()

Flight_ID           0
Aeronave            0
DepartureStation    0
ArrivalStation      0
Destination_Type    0
Origin_Type         0
STD                 0
STA                 0
Capacity            0
Passengers          0
Bookings            0
dtype: int64

In [11]:
# Ordenar por fecha
flights = flights.sort_values(by="STD")

Se crea una columna de `Semana` y `Mes` para análisis posteriores.

In [12]:

flights["Mes"] = flights["STD"].dt.month
flights["Semana"] = flights["STD"].dt.isocalendar().week
flights["DiaSemana"] = flights["STD"].dt.weekday + 1
flights["Hora"] = flights["STD"].dt.hour
flights

Flight_ID Aeronave DepartureStation  \
6696    030f35839975b1721e2427db425ef402   XA-VAJ               BM   
178553  7b9f51daca0f560676d28aee3d5666f7   XA-VAW               BM   
53258   d4ac6f26a94efea00002147f15d96a41   XA-VIU               BM   
123223  555496d4217768d1cc276ffbe83235d5   XA-VXF               AT   
53248   01cb03565042d43a1deafae09eef0ca1   XA-VAN               BM   
...                                  ...      ...              ...   
221868  a665d50e95bf905d862563facacbf05a   XA-VXA               AO   
92561   6dda5bd834419dd5b900408b0f5fd77b   XA-VAU               BM   
105522  c711d31ee17b65d788d2df7ebde346f1   XA-VBI               AK   
57082   dc6b4fa09794681fbf19cdd98f5a942e   XA-VIF               BM   
102919  34771f9eadb3f935f45d3271fa9929ce   XA-VXP               BM   

       ArrivalStation     Destination_Type        Origin_Type  \
6696               AV  MX Amigos y Familia  Ciudad Fronteriza   
178553             BA     Ciudad Principal  Ciudad Fronteriza   
53258              AO     Ciudad Principal  Ciudad Fronteriza   
123223             AU           Ecoturismo   Ciudad Principal   
53248              AO     Ciudad Principal  Ciudad Fronteriza   
...               ...                  ...                ...   
221868             BM    Ciudad Fronteriza   Ciudad Principal   
92561              AT     Ciudad Principal  Ciudad Fronteriza   
105522             AT     Ciudad Principal              Playa   
57082              AO     Ciudad Principal  Ciudad Fronteriza   
102919             AT     Ciudad Principal  Ciudad Fronteriza   

                       STD                 STA  Capacity  Passengers  \
6696   2023-01-02 00:20:00 2023-01-02 05:20:00       186       120.0   
178553 2023-01-02 00:30:00 2023-01-02 06:00:00       186       170.0   
53258  2023-01-02 00:35:00 2023-01-02 05:30:00       186       185.0   
123223 2023-01-02 00:55:00 2023-01-02 02:45:00       240       168.0   
53248  2023-01-02 01:15:00 2023-01-02 06:10:00       186       165.0   
...                    ...                 ...       ...         ...   
221868 2023-12-31 23:00:00 2024-01-01 00:05:00       240       201.0   
92561  2023-12-31 23:05:00 2024-01-01 04:30:00       186       131.0   
105522 2023-12-31 23:15:00 2024-01-01 00:45:00       220       217.0   
57082  2023-12-31 23:45:00 2024-01-01 04:40:00       186       168.0   
102919 2023-12-31 23:50:00 2024-01-01 05:15:00       240       161.0   

        Bookings  Mes  Semana  DiaSemana  Hora  
6696        68.0    1       1          1     0  
178553      99.0    1       1          1     0  
53258       89.0    1       1          1     0  
123223     118.0    1       1          1     0  
53248       92.0    1       1          1     1  
...          ...  ...     ...        ...   ...  
221868      90.0   12      52          7    23  
92561       81.0   12      52          7    23  
105522     126.0   12      52          7    23  
57082       75.0   12      52          7    23  
102919     116.0   12      52          7    23  

[121724 rows x 15 columns]

In [13]:
flights["HourDuration"] = (flights["STA"] - flights["STD"]).dt.total_seconds() / 3600
flights.head()

Flight_ID Aeronave DepartureStation  \
6696    030f35839975b1721e2427db425ef402   XA-VAJ               BM   
178553  7b9f51daca0f560676d28aee3d5666f7   XA-VAW               BM   
53258   d4ac6f26a94efea00002147f15d96a41   XA-VIU               BM   
123223  555496d4217768d1cc276ffbe83235d5   XA-VXF               AT   
53248   01cb03565042d43a1deafae09eef0ca1   XA-VAN               BM   

       ArrivalStation     Destination_Type        Origin_Type  \
6696               AV  MX Amigos y Familia  Ciudad Fronteriza   
178553             BA     Ciudad Principal  Ciudad Fronteriza   
53258              AO     Ciudad Principal  Ciudad Fronteriza   
123223             AU           Ecoturismo   Ciudad Principal   
53248              AO     Ciudad Principal  Ciudad Fronteriza   

                       STD                 STA  Capacity  Passengers  \
6696   2023-01-02 00:20:00 2023-01-02 05:20:00       186       120.0   
178553 2023-01-02 00:30:00 2023-01-02 06:00:00       186       170.0   
53258  2023-01-02 00:35:00 2023-01-02 05:30:00       186       185.0   
123223 2023-01-02 00:55:00 2023-01-02 02:45:00       240       168.0   
53248  2023-01-02 01:15:00 2023-01-02 06:10:00       186       165.0   

        Bookings  Mes  Semana  DiaSemana  Hora  HourDuration  
6696        68.0    1       1          1     0      5.000000  
178553      99.0    1       1          1     0      5.500000  
53258       89.0    1       1          1     0      4.916667  
123223     118.0    1       1          1     0      1.833333  
53248       92.0    1       1          1     1      4.916667

In [14]:
flights["HourDuration"].describe()

count    121724.000000
mean          2.052066
std           1.015633
min           0.250000
25%           1.416667
50%           1.666667
75%           2.500000
max          27.333333
Name: HourDuration, dtype: float64

In [15]:
(flights["HourDuration"] > 5).sum()

1500

Se incluyen vuelos de itinerario sumados como hasta 27+ horas de viaje. Esto parece un error en los datos, por lo que se eliminan estos registros los cuales únicamente representan el 1500 de alrededor de 121,724 de los datos.

In [16]:
flights = flights[flights["HourDuration"] <= 5]
flights

Flight_ID Aeronave DepartureStation  \
6696    030f35839975b1721e2427db425ef402   XA-VAJ               BM   
53258   d4ac6f26a94efea00002147f15d96a41   XA-VIU               BM   
123223  555496d4217768d1cc276ffbe83235d5   XA-VXF               AT   
53248   01cb03565042d43a1deafae09eef0ca1   XA-VAN               BM   
99546   f0736ca7a01a15bb76b2491d3108004f   XA-VXF               AU   
...                                  ...      ...              ...   
227849  20f799384c5f51760c99f62bf40b2db5   XA-VIP               BA   
220493  2cca4ce62f73477f7a113e83f0760074   XA-VAA               AD   
221868  a665d50e95bf905d862563facacbf05a   XA-VXA               AO   
105522  c711d31ee17b65d788d2df7ebde346f1   XA-VBI               AK   
57082   dc6b4fa09794681fbf19cdd98f5a942e   XA-VIF               BM   

       ArrivalStation     Destination_Type          Origin_Type  \
6696               AV  MX Amigos y Familia    Ciudad Fronteriza   
53258              AO     Ciudad Principal    Ciudad Fronteriza   
123223             AU           Ecoturismo     Ciudad Principal   
53248              AO     Ciudad Principal    Ciudad Fronteriza   
99546              AT     Ciudad Principal           Ecoturismo   
...               ...                  ...                  ...   
227849             BM    Ciudad Fronteriza     Ciudad Principal   
220493             BM    Ciudad Fronteriza  MX Amigos y Familia   
221868             BM    Ciudad Fronteriza     Ciudad Principal   
105522             AT     Ciudad Principal                Playa   
57082              AO     Ciudad Principal    Ciudad Fronteriza   

                       STD                 STA  Capacity  Passengers  \
6696   2023-01-02 00:20:00 2023-01-02 05:20:00       186       120.0   
53258  2023-01-02 00:35:00 2023-01-02 05:30:00       186       185.0   
123223 2023-01-02 00:55:00 2023-01-02 02:45:00       240       168.0   
53248  2023-01-02 01:15:00 2023-01-02 06:10:00       186       165.0   
99546  2023-01-02 03:25:00 2023-01-02 05:30:00       240       161.0   
...                    ...                 ...       ...         ...   
227849 2023-12-31 22:40:00 2024-01-01 00:20:00       186       141.0   
220493 2023-12-31 22:40:00 2023-12-31 23:55:00       186       181.0   
221868 2023-12-31 23:00:00 2024-01-01 00:05:00       240       201.0   
105522 2023-12-31 23:15:00 2024-01-01 00:45:00       220       217.0   
57082  2023-12-31 23:45:00 2024-01-01 04:40:00       186       168.0   

        Bookings  Mes  Semana  DiaSemana  Hora  HourDuration  
6696        68.0    1       1          1     0      5.000000  
53258       89.0    1       1          1     0      4.916667  
123223     118.0    1       1          1     0      1.833333  
53248       92.0    1       1          1     1      4.916667  
99546       90.0    1       1          1     3      2.083333  
...          ...  ...     ...        ...   ...           ...  
227849      75.0   12      52          7    22      1.666667  
220493      88.0   12      52          7    22      1.250000  
221868      90.0   12      52          7    23      1.083333  
105522     126.0   12      52          7    23      1.500000  
57082       75.0   12      52          7    23      4.916667  

[120224 rows x 16 columns]

In [17]:
flights.to_parquet("../resources/flightstest2_clean.parquet", index=False)

## Ventas

In [18]:
sales = pd.read_parquet("../resources/sales.parquet")

In [19]:
sales.head()

Flight_ID ProductType  \
0  a05290288259526edd3601160b10e1de     Botanas   
1  08f6f97437df8db101b050f1110be656     Licores   
2  4e09c949826a77207868412baeff6d30     Licores   
3  c3e1568fe46c68d3174681d322d412b3     Licores   
4  7396e43fb94caca87add7bb1f1611e7e     Vivabus   

                      ProductName  Quantity  TotalSales  
0             Carne Seca Habanero         1        55.0  
1                   Jw Red Label          2       240.0  
2                   Jack And Coke         8       576.0  
3                    Jw Red Label         2       240.0  
4  Vivabus_Gdl.-Nvo. C. Camionera         4       320.0

In [20]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2564361 entries, 0 to 2564360
Data columns (total 5 columns):
 #   Column       Dtype  
---  ------       -----  
 0   Flight_ID    object 
 1   ProductType  object 
 2   ProductName  object 
 3   Quantity     int64  
 4   TotalSales   float64
dtypes: float64(1), int64(1), object(3)
memory usage: 97.8+ MB


In [21]:
sales.describe()

Quantity    TotalSales
count  2.564361e+06  2.564361e+06
mean   2.519319e+00  1.110370e+02
std    2.829032e+00  1.736604e+02
min    1.000000e+00 -3.234000e+03
25%    1.000000e+00  4.500000e+01
50%    2.000000e+00  8.000000e+01
75%    3.000000e+00  1.440000e+02
max    8.500000e+01  1.290000e+04

In [22]:
sales["Flight_ID"].value_counts().head(10)

Flight_ID
09a295c868301fe15df13d8411cd4d37    73
4c65ba84403c407f59258b9ad414e40d    73
a44008629d97a1b4c9b6e0980ae8b11a    73
edbdab3c1c0df3d2b2213343fb9144eb    73
b6a39758032c87d2e315889e36b97277    72
14069ac4ad5e26456a8f103484b84774    71
ab6dc691f1869a7c60dfd306ea408cdf    71
fd0e971f1c0af6e01be0cec64326d09b    70
c09524fd81d49de6643b61b31bae4864    69
722a5e63351b62bc5e95488f77f1c84e    69
Name: count, dtype: int64

In [23]:
# Se eliminan valores nulos pero
# realmente no hay
sales.dropna(inplace=True)

In [24]:
sales["ProductType"].value_counts()

ProductType
Botanas                 613793
Refrescos               585118
Specials                395760
Licores                 304987
Perecederos             181990
Galletas                143928
Bebidas Calientes       125235
Sopas                   115486
OFERTAS                  38384
Lacteos                  37755
Vivabus                   9819
Transportaciones CUN      3635
Hertz.                    2398
COMBOS CREW               2017
VIVA Taxis                1384
Transportaciones MTY      1220
VivaTransfer               435
Antros                     388
VIVA PLAY                  386
Transportaciones TLC       239
Alimentos Charter            4
Name: count, dtype: int64

Se eliminan aquellos tipos de producto que no son de interés para el análisis pues no son tangibles

In [25]:
excluded_types = [
    "Specials",
    "OFERTAS ",
    "Vivabus",
    "Transportaciones CUN",
    "Hertz.",
    "COMBOS CREW",
    "VIVA Taxis",
    "Transportaciones MTY",
    "VivaTransfer",
    "Antros",
    "VIVA PLAY",
    "Transportaciones TLC",
    "Alimentos Charter",
]

sales = sales[~sales.ProductType.isin(excluded_types)]
sales["ProductType"].value_counts()

ProductType
Botanas              613793
Refrescos            585118
Licores              304987
Perecederos          181990
Galletas             143928
Bebidas Calientes    125235
Sopas                115486
Lacteos               37755
Name: count, dtype: int64

In [26]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2108292 entries, 0 to 2564360
Data columns (total 5 columns):
 #   Column       Dtype  
---  ------       -----  
 0   Flight_ID    object 
 1   ProductType  object 
 2   ProductName  object 
 3   Quantity     int64  
 4   TotalSales   float64
dtypes: float64(1), int64(1), object(3)
memory usage: 96.5+ MB


Se eliminan aquellas ventas que sean negaivas pues implican descuentos

In [ ]:
sales = sales[sales.TotalSales > 0]

In [ ]:
sales.drop(columns=["TotalSales"], inplace=True)

In [ ]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2107861 entries, 0 to 2564360
Data columns (total 4 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   Flight_ID    object
 1   ProductType  object
 2   ProductName  object
 3   Quantity     int64 
dtypes: int64(1), object(3)
memory usage: 80.4+ MB


In [ ]:
sales.reset_index(inplace=True)

In [ ]:
# Usar nuevos datos de sales
sales.to_parquet("../resources/sales_cleaned.parquet")

c:\Users\Franco Mendoza\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarrow\pandas_compat.py:373: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if _pandas_api.is_sparse(col):
